lightning-invoice/src/utils.rs

```diff
@@ -381,39 +405,57 @@ where
 ///   `is_usable` (i.e. the peer is connected).
 /// * If any public channel exists, the returned `RouteHint`s will be empty, and the sender will
 ///   need to find the path by looking at the public channels instead
-fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Option<u64>) -> Vec<RouteHint>{
+fn filter_channels<L: Deref>(
+	channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Option<u64>, logger: &L
+) -> Vec<RouteHint> where L::Target: Logger {
 	let mut filtered_channels: HashMap<PublicKey, ChannelDetails> = HashMap::new();
 	let min_inbound_capacity = min_inbound_capacity_msat.unwrap_or(0);
 	let mut min_capacity_channel_exists = false;
 	let mut online_channel_exists = false;
 	let mut online_min_capacity_channel_exists = false;
 
+	log_trace!(logger, "Considering {} channels for invoice route hints", channels.len());
 	for channel in channels.into_iter().filter(|chan| chan.is_channel_ready) {
 		if channel.get_inbound_payment_scid().is_none() || channel.counterparty.forwarding_info.is_none() {
+			log_trace!(logger, "Ignoring channel {} for invoice route hints", log_bytes!(channel.channel_id));
 			continue;
 		}
 
 		if channel.is_public {
 			// If any public channel exists, return no hints and let the sender
 			// look at the public channels instead.
+			log_trace!(logger, "Not including channels in invoice route hints on account of public channel {}",
+				log_bytes!(channel.channel_id));
 			return vec![]
 		}
 
 		if channel.inbound_capacity_msat >= min_inbound_capacity {
-			min_capacity_channel_exists = true;
+			if !min_capacity_channel_exists {
+				log_trace!(logger, "Channel with enough inbound capacity exists for invoice route hints");
+				min_capacity_channel_exists = true;
+			}
+
 			if channel.is_usable {
 				online_min_capacity_channel_exists = true;
 			}
 		}
+
 		if channel.is_usable {
-			online_channel_exists = true;
+			if !online_channel_exists {
+				log_trace!(logger, "Channel with connected peer exists for invoice route hints");
+				online_channel_exists = true;
+			}
 		}
+
 		match filtered_channels.entry(channel.counterparty.node_id) {
 			hash_map::Entry::Occupied(mut entry) => {
 				let current_max_capacity = entry.get().inbound_capacity_msat;
 				if channel.inbound_capacity_msat < current_max_capacity {
 					continue;
 				}
+				log_trace!(logger, "Preferring channel {} ({} msats) over {} ({} msats) for invoice route hints",
```

Comment:
Could also be worth mentioning the counterparty node id for the channel here, due to the filtering only including 1 channel per counterparty node.